### PR TITLE
Fix Bug with setting Colour

### PR DESCRIPTION
**Pull request by Wilma456**
If you calls term.setBackgroundColor or term.setTextColor with the wrong argument, your computer will crash. With my Bugfix only your program crash and your computer keep runing.

### CODE REVIEW

**Comment by Wilma456** [comment]
Try term.setBackgroundColor(2000). This will crashed your Computer and it's a number.  The Code from SquidDev doesen't stop this.

**Comment by SquidDev** [comment]
@Wilma456 That seems a bug with multishell: `term.native().setBackgroundColour(2000)` works as expected.

**Comment by Wilma456** [comment]
The Code now checks just, if color is a number. The most errors are caused by spell errors, so it will solve the most errors. If it a bug in multishell, it must be fixed in multishell I think.

File: src/main/resources/assets/computercraft/lua/rom/apis/window [hunk 1]
```diff
@@ -270,19 +270,19 @@ function create( parent, nX, nY, nWidth, nHeight, bStartVisible )
...
-        nTextColor = color
-        if bVisible then
-            updateCursorColor()
+        if type(color) == "number" then
+	    nTextColor = color
+            if bVisible then
+               updateCursorColor()
+            end
+        else
+           error("Color must be a number")
```
Comment (SquidDev) [inline src/main/resources/assets/computercraft/lua/rom/apis/window]: Would probably be better to use `error("Expected number", 2)` here, as that is what the `term` API uses.

**Comment by SquidDev** [comment]
> If it a bug in multishell, it must be fixed in multishell I think.

Sorry, it is a bug in the window API. I always conflate the two for whatever reason. My suggested change involving `math.log` *should* solve the problem, but is rather inefficient.

**Comment by Wilma456** [comment]
nTextColor = math.log(color) and nTextColor = math.floor(math.log(color) / math.log(2)) + 1 does not work. The computer crashed on boot.

**Comment by SquidDev** [comment]
@Wilma456 Did you change the `tHex` table too? I'll have a look - I'd have thought it would work.

**Comment by Wilma456** [comment]
No. What I should change in the table?

**Comment by SquidDev** [comment]
Right. Having another look at the code, I think the "best" solution would be:

You basically normalise the colour to be between 1 and 16, so it needs to change to a simple array:

```lua
local tHex = { "0", "1", "2", -- [[ etc to "f" ]] }
```
An alternative solution would be to have both. So do:

```lua
local floor, log, log2 = math.floor, math.log, math.log(2)
local function setBackgroundColor( color )
    -- Type assertions, whatever, etc...
    if not tHex[color] then
        if color < 0 or color > 0xFFFF then error(""Colour out of range", 2) end
         color = 2^floor(log(color) / log2)
    end
    nBackgroundColor = color
end
```

**Comment by Wilma456** [comment]
@SquidDev With this code the colours are wrong

**Comment by SquidDev** [comment]
@Wilma456 Can you elaborate? It appears to work correctly for me.

**Comment by Wilma456** [comment]
@SquidDev 

Here's my actual code: https://pastebin.com/qRmyUevL

Ans here's a picture of CraftOS using this code: http://www.bilder-upload.eu/show.php?file=67ed09-1494688304.png

**Comment by SquidDev** [comment]
Your implementation of `setTextColor` sets `nBackgroundColor` instead of `nTextColor`.

**Comment by Wilma456** [comment]
@SquidDev Copy and paste Error. Sorry. :man_facepalming: 

@dan200 The bug is now fixed and the PR is ready to merge.

File: src/main/resources/assets/computercraft/lua/rom/apis/window [hunk 2]
```diff
@@ -268,16 +268,19 @@ function create( parent, nX, nY, nWidth, nHeight, bStartVisible )
...
     local function setTextColor( color )
-        if type(color) == "number" then
-	    nTextColor = color
-            if bVisible then
-               updateCursorColor()
-            end
-        else
-           error("Color must be a number")
-        end
+	if not(type(color) == "number") then
```
Comment (SquidDev) [inline src/main/resources/assets/computercraft/lua/rom/apis/window]: It'd probably be nicer to use `~=` here - `if type(color) ~= "number" then`.

Comment (SquidDev) [inline src/main/resources/assets/computercraft/lua/rom/apis/window]: Also, not entirely sure what is going on with the indentation here - it seems like a really weird mix of spaces and tabs.

**Comment by dan200** [comment]
This code is very messy. Does someone want to just submit a cleaner version of this same code that just checks if the number is a key in tHex? We don't need to support people passing in non-colours.

**Comment by Wilma456** [comment]
@dan200 Done!

File: src/main/resources/assets/computercraft/lua/rom/apis/window [hunk 3]
```diff
@@ -270,19 +270,23 @@ function create( parent, nX, nY, nWidth, nHeight, bStartVisible )
...
+       if type(tHex[color]) == "string" then
         nTextColor = color
         if bVisible then
             updateCursorColor()
         end
+       else
+        error("Colour out of range", 2)
+       end
+    else
+       error("Color must be a number",2)
```
Comment (KnightMiner) [inline src/main/resources/assets/computercraft/lua/rom/apis/window]: Minor issue, but this line spells it "color" and the above is "colour", being consistent with whatever old errors do would be better.

File: src/main/resources/assets/computercraft/lua/rom/apis/window [hunk 4]
```diff
@@ -312,17 +316,21 @@ function create( parent, nX, nY, nWidth, nHeight, bStartVisible )
...
 
     local function setBackgroundColor( color )
+    if type(color) == "number" then
+       if type(tHex[color]) == "string" then
         nBackgroundColor = color
+       else
+        error("Colour out of range",2)
+       end
+    else
+       error("Color must be a number",2)
```
Comment (KnightMiner) [inline src/main/resources/assets/computercraft/lua/rom/apis/window]: Same here

File: src/main/resources/assets/computercraft/lua/rom/apis/window [hunk 5]
```diff
@@ -312,17 +316,21 @@ function create( parent, nX, nY, nWidth, nHeight, bStartVisible )
...
-    function window.setBackgroundColor( color )
-        setBackgroundColor( color )
     end
 
-    function window.setBackgroundColour( color )
-        setBackgroundColor( color )
-    end
+    window.setBackgroundColor = setBackgroundColor
 
+    window.setBackgroundColour = setBackgroundColo
```
Comment (KnightMiner) [inline src/main/resources/assets/computercraft/lua/rom/apis/window]: Should be "setBackgroundColor", not "setBackgroundColo"

File: src/main/resources/assets/computercraft/lua/rom/apis/window [hunk 6]
```diff
@@ -270,19 +270,23 @@ function create( parent, nX, nY, nWidth, nHeight, bStartVisible )
     end
 
     local function setTextColor( color )
+	if type(color) == "number" then
+       if type(tHex[color]) == "string" then
```
Comment (KnightMiner) [inline src/main/resources/assets/computercraft/lua/rom/apis/window]: Indentation seems a little odd here

**Comment by dan200** [comment]
This code is still super messy. Rather than having a long conversation spelling out how to fix, i've just committed a cleaner implementation: https://github.com/dan200/ComputerCraft/commit/8c4331d15aaf07b87c1033f36799813bf633a188
Closing as this merge is no longer necessary

**Pull request closed without merging**